design/object-store-user.md

```diff
@@ -0,0 +1,131 @@
+1. Remove the restriction that a CephObjectStoreUser be located in the same namespace as the ObjectStore.
+1. Change the ownerReference of the generated key-pair Secret to reference the CephObjectStoreUser rather than the ceph-cluster.
+1. Change the name of the generated secret per [1] above. Note: this change is also **not** backwards compatible.
+
+#### Storage Class
+A [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/) is a global Kubernetes resource with the ability to pass parameters to a provisioner. Storage classes further abstract the underlying storage based on admin policies. Storage class visibility can be controlled by RBAC rules. A simple storage class for a rook-ceph object store might look like this:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-ceph-object-store
+provisioner: kubernetes.io/rook-ceph-operator
+parameters:
+  objectStore: rook-ceph/objectStoreName  # format: “namespace/objectStoreName”
+```
+There should be a 1:1 mapping of storage class and rook object store. For example, if there are 5 object stores, each in their own namespace, then 5 global storage classes are needed. Likewise, if there are 5 object stores in a single namespace, then, again, 5 storage classes are required. Note that the creation of an object store does not require an associated storage class. Using storage class more closely adheres to the Kubernetes storage model and offers greater flexibility to the developer and storage admin.
+
+#### Namespace
+When a storage class is specified the object store’s namespace is found in the storage class parameter. A typical object store user definition might look like this:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: cephObjectStoreUser
+metadata:
+  name: my-user-1
+    namespace: dev-user
+spec:
+  storageClass: rook-ceph-object-store  # name of storage class
+  displayName: aUserDisplayName
+```
+Through the `storageClass` value we know the namespace and name of the object store.
+
+#### Secret OwnerReference
+Today, the generated key-pair secret’s owner reference is the id of the ceph cluster, which is also the same owner set for the CephObjectStoreUser. Thus, using the Kubernete’s [cascading deletes feature](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion), the secret is not actually deleted until the ceph cluster is deleted. This enhancement proposes to set the Secret’s OwnerReference to the id of the CephObjectStoreUser that created the secret. So, when the CephObjectStoreUser is deleted, its secret is automatically deleted. A finalizer will also be added so that the the generated secret cannot be deleted until its ownerCephObjectStoreUser is first deleted.
```

Comment:
Why is the finalizer needed? The ownerReference should already take care of deleting the resources in the proper order. Is there some reason to veto deleting the secret? The only place we currently have a finalizer in a crd is for the cluster to veto the deletion until all the mounts are cleaned up.

Reply:
Our understanding is that adding our own finalizer prevents the generated secret from being deleted before the user is deleted. IOW, prevent the user from using `kubectl delete <secret>` to delete only the secret.

Reply:
@travisn thanks for your comments. What does ceph/ragosgw do when a bucket's owner's secret is deleted? Can that bucket still be managed within ceph?

Reply:
> Is the purpose so the user doesn't shoot themselves in the foot and delete the secret accidentally? 

Yes. Though this might not be desired. I see 3 scenarios around prematurely deleting the secret.
1. just delete the secret per the user's desire. How is this handled by ceph/s3?
1. fail the deletion of the secret (perhaps via an operator-added finalizer)
1. delete the secret and code the operator to notice the secret is missing and regenerate the secret. Whap happens with to bucket access when new key-pairs are used?

Reply:
As long as the radosgw user exists, credentials can be requested for that user. The creds that are saved in the K8s secret can't be regenerated, but another set of creds can be generated for the same user. Similarly, in AWS you can generate multiple sets of creds for the same S3 user.

But if the radosgw user is deleted, I'm not sure the behavior. Perhaps the admin can re-assign the bucket owner. @yehudasa can you comment on what happens when a user is deleted when they are the bucket owner?